### ISSUE TITLE
Canonicalisation algorithm

### ISSUE THREAD

**Issue by arnau** [openregister/specification#75]
The canonicalisation algorithm is described as follows:

>JSON object values MUST be sorted into lexicographical order. The keys of a JSON object must be a valid field name, which is restricted to the alphabet of lower case letters and hyphens, which makes this ordering relatively simple to implement.
>
>All whitespace MUST be removed.
>
>Characters in strings must be represented as follows:
>
>For ASCII control characters (codepoints 0x00 - 0x1f):
>
>If it has a short representation (\b, \f, \n, \r, or \t), that short representation MUST be used.
>
>Other control characters (such as NULL) MUST be represented as a \u00XX escape sequence. Hexadecimal digits MUST be upper-case.
>
>Backslash (\) and double quote (") MUST be escaped as \\ and \" respectively.
>
>All other characters MUST be included literally (ie unescaped). This includes forward-slash (/).
>
>This canonicalisation algorithm is very similar to that used in [JCS], except that we stipulate an ordering of keys, and we enforce upper-case rather than lower-case hex digits.

The escape section for values can be removed entirely given that both json.org and RFC7159 require the same escape approach. With two exceptions, the requirement for representing hex values as uppercase and forcing solidus (U+002F) to be unescaped and non-control characters must be unescaped as well. So a rewrite could be:

> * The blob MUST be a valid JSON object according to RFC7159.
> * All insignificant whitespace according to RFC7159 MUST be removed.
> * The JSON object keys must be a valid field name, which is restricted to the alphabet of lower case letters and hyphens (`[a-z-]`).
> * The JSON object values MUST be sorted into lexicographical order.
> * Unicode sequences `\uXXXX` MUST be in upper-case.
> * The forward slash or solidus (`/`) MUST be unescaped.
> * Non-control characters (i.e. out of the range `\u0000 .. \u001F`) MUST be unescaped.

**Comment by benlaurie** [comment]
I don't believe it is true that field names must be lower case according to the JSON spec.

**Comment by arnau** [comment]
JSON objects can indeed have upper case characters. What ths list tries to say is that, on top of being a valid JSON key, it must conform to `[a-z-]` (so no upper case, no digits, etc. I'll try to rephrase it to make it more clear.

**Comment by arnau** [comment]
Addressed in the `next` branch and obsoleted by https://github.com/openregister/registers-rfcs/pull/24 (when approved).
